### PR TITLE
js-oop-frogger

### PR DESCRIPTION
**Pull request by justdevway**
_Complete the form below and remove this line._

### js-oop-frogger

[Demo](https://justdevway.github.io/frogger/) |
[Code base](https://github.com/justdevway/frontend-2019-homeworks/tree/js-oop-frogger/submissions/justdevway/js-oop-frogger)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@justdevway Thank you for the submission.

Please, check yourself against the following points:
 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as constants
   * [ ] there are core contants and derived constants
         (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
 * Requirements re **OOP**:
   * [ ] properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
 * Most common mistakes
   * [ ] `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
         `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
         [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)
         
Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by justdevway** [comment]
@OleksiyRudenko, thanks for this detail feedback :) 
Can your answer for my questions below:
1) So, sometimes I can use ES6 or use only for variables ? 
2) All number  I need convert to variable ? 
3) Try to use maximum correct OOP ?

Thanks

**Comment by OleksiyRudenko** [comment]
> @OleksiyRudenko, thanks for this detail feedback :)
> Can your answer for my questions below:
> 
> 1. So, sometimes I can use ES6 or use only for variables ?
You can use ES6 to its full extent with the only exclusion: use prototype object model, not ES6 classes syntax.

> 2. All number  I need convert to variable ?
Either `var` or preferably `const`. In either case all caps name will denote these are constants by semantic meaning.

> 3. Try to use maximum correct OOP ?
Yes. Use OOP, be SOLID, keep code DRY

> 
> Thanks

**Comment by justdevway** [comment]
Okay, thanks! 
According to plan - do it today! :)

File: submissions/justdevway/js-oop-frogger/js/app.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+};
+
+
+
+
+var player = new Player(player_parametrs.x(), player_parametrs.y(), player_parametrs.sprite);
+var enemy1 =  new Enemy(enemy_parametrs.x, enemy_parametrs.y[0], enemy_parametrs.speed(), enemy_parametrs.sprite);
+var enemy2 =  new Enemy(enemy_parametrs.x, enemy_parametrs.y[1], enemy_parametrs.speed(), enemy_parametrs.sprite);
+var enemy3 =  new Enemy(enemy_parametrs.x, enemy_parametrs.y[2], enemy_parametrs.speed(), enemy_parametrs.sprite);
+var allEnemies = [enemy1, enemy2, enemy3];
```
Comment (OleksiyRudenko) [inline submissions/justdevway/js-oop-frogger/js/app.js]: Can be easily optimized like
```
var allEnemies = [0,1,2].map(index => new Enemy(..., enemy_parameters.y[index], ...);
```

File: submissions/justdevway/js-oop-frogger/js/app.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        'images/princess.png',
+        'images/Key.png'
+    ]
+};
+var bonus_parametrs = {
+    x: () => Math.floor(Math.random() * CANVAS_WIDTH  * 0.6 + CANVAS_WIDTH * 0.1),
+    y: () => Math.floor(Math.random() * CANVAS_HEIGHT * 0.3 + CANVAS_HEIGHT * 0.25),
+    sprite: 'images/Heart.png'
+};
+
```
Comment (OleksiyRudenko) [inline submissions/justdevway/js-oop-frogger/js/app.js]: `*_parameters` are constants by their nature. Convention is to either have those all caps (`ENEMY_PARAMETERS`) or camel case like regular variables (`enemyParameters`).

Comment (justdevway) [reply submissions/justdevway/js-oop-frogger/js/app.js]: Okay, I will make this changes

**Pull request merged**